### DISCUSSION TITLE
How can I fetch with joins using "relationLoadStrategy"

### DISCUSSION THREAD

**Discussion by nix0x00** [prisma/prisma#23395] [Q&A] [answered]
### Question

I am fetching data from departments based on the user's interactions by using the raw query mentioned below. Now that prisma has launched relations feature, how can I actually utilize it and can following query be converted to prisma query so that I don't have to use raw query.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```  this.prismaService.$queryRawUnsafe<any[]>(
      `
      SELECT D.*, COUNT(I.department_id) AS interaction_count
      FROM Department D
      LEFT JOIN Interactions I ON D.id = I.department_id AND I.user_id = ?
      GROUP BY D.id
      ORDER BY interaction_count DESC;
      `,
      userId,
    );
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment, marked as answer]
Hi @nix0x00 👋 

Thank you for raising this question.

> Now that prisma has launched relations feature, how can I actually utilize it.

To use joins, you will need to enable the new `relationLoadStrategy`, you'll first need to add the preview feature flag to the generator block of your Prisma Client:
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["relationJoins"]
}
```
Once that's done, you'll need to re-run prisma generate for this change to take effect and pick a relation load strategy in your queries.
Here is an example that uses the new join strategy:
```js
const usersWithPosts = await prisma.user.findMany({
  relationLoadStrategy: "join", // or "query"
  include: {
    posts: true,
  },
});
```
You should note that you can choose between database level join and application level join. With database level join, you would use `join` as the value for the `relationLoadStrategy` and with application level join, you would use `query`. You can take a look at this blog [post](https://www.prisma.io/blog/prisma-orm-now-lets-you-choose-the-best-join-strategy-preview) to learn more.

> can following query be converted to prisma query so that I don't have to use raw query.

If you want to convert this raw SQL query into a Prisma query, you might need to use the `groupBy` method provided by Prisma. But, as of now, Prisma does not support grouping on relations or counting based on a relation, as mentioned in the [Github discussion](https://github.com/prisma/prisma/discussions/19857). So at the moment, it's not currently possible to convert your raw SQL query into a Prisma query due to the limitations of the `groupBy` method.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by janpio** [reply]
Note: Enabling the `relationJoins` preview feature already flips the default to `relationLoadStrategy: "join"`, so you will not have to define that for every Prisma Client query.

**Reply by nix0x00** [reply]
Got it, thank you. I think this is a must feature that prisma should have in near future. How and where can I suggest this feature so Prisma team can at least think about it.

**Reply by janpio** [reply]
Which feature exactly? There is most probably already a feature request issue for it, so search those first.
